api/http-api.md

```diff
@@ -0,0 +1,177 @@
+* Task polling/completing:
+  * Intentionally not exposed (except for activity task stuff needed for async activities)
+* Task queues:
+  * `ResetStickyTaskQueue` - Intentionally not exposed
+  * `DescribeTaskQueue` - `GET /api/v1/namespaces/{namespace}/task-queues/{task_queue_name}`
+  * `ListTaskQueuePartitions` - Intentionally not exposed
+* Schedules:
+  * `CreateSchedule` - `POST /api/v1/namespaces/{namespace}/schedules/{schedule_id}`
+  * `DescribeSchedule` - `GET /api/v1/namespaces/{namespace}/schedules/{schedule_id}`
+  * `UpdateSchedule` - `POST /api/v1/namespaces/{namespace}/schedules/{schedule_id}/update`
+    * This is not a `PUT` because update different from create and not `PATCH` to avoid confusion with below
+  * `PatchSchedule` - `POST /api/v1/namespaces/{namespace}/schedules/{schedule_id}/patch`
+    * This is not a `PATCH` to avoid confusion with update above
+  * `ListScheduleMatchingTimes` - `GET /api/v1/namespaces/{namespace}/schedules/{schedule_id}/matching-times`
+  * `DeleteSchedule` - `DELETE /api/v1/namespaces/{namespace}/schedules/{schedule_id}`
+    * Confirm whether eventually consistent
+  * `ListSchedules` - `GET /api/v1/namespaces/{namespace}/schedules`
+* Batch operations:
+  * `StartBatchOperation` - `POST /api/v1/namespaces/{namespace}/batch-operations/{job_id}`
+  * `DescribeBatchOperation` - `GET /api/v1/namespaces/{namespace}/batch-operations/{job_id}`
+  * `ListBatchOperations` - `GET /api/v1/namespaces/{namespace}/batch-operations`
+  * `StopBatchOperation` - `POST /api/v1/namespaces/{namespace}/batch-operations/{job_id}/stop`
+* Worker build IDs:
+  * Intentionally not exposed until they settle
+* Other calls:
+  * `GetClusterInfo` - Intentionally not exposed
+  * `GetSystemInfo` - `GET /api/v1/system-info`
+
+### Embedding into Server
+
+Current suggested approach, subject to discussion:
+
+* Generated gRPC gateway code will be auto-generated at the same time as api-go is for gRPC and live alongside it
+* Service config has new `httpPort`
+  * Only allowed on "frontend" service, presence elsewhere is config failure
+  * Default value is `grpcPort` + 10
+    * Makes `7243` the default - unassigned in IANA, unused in server
+    * Cannot think of case where +10 clashes, e.g. if they use 80 or 443 or 8080 or whatever, 90, 453, and 8090 are fine
+* Service config has new `disableHttp`
+  * Default enabled, as harmless as exposing gRPC by default, and will encourage use
+  * This is separate from `httpPort` because we don't want to support some `false` on an int option
+* Full support for same TLS options as frontend
+* Since it goes through gRPC, all metadata (i.e. HTTP headers) and interceptors are respected
+* Will test all of the above
+* Can discuss approach for cloud in separate place
+
+### Documentation
+
+* Need to simply document its presence at first, OpenAPI spec _can_ come later though Temporal may consider a more
+  intentionally-designed API in the future
```

Comment:
Can the OpenAPI spec be generated automatically? I think it might be a bit unfriendly to provide the API, tell people it exists, and then not document how to use it.

Reply:
It _can_ be generated automatically. I am not making it a part of the initial project because OpenAPI sucks and maintaining it sucks and there are a lot of little pieces we may want to add based on auth, docs, etc. I don't think the project needs that at first.

> I think it might be a bit unfriendly to provide the API, tell people it exists, and then not document how to use it.

No more unfriendly than what happens today. I think we should provide the API first, we can discuss the rest later.

Reply:
Yeah that's fair. I was thinking about this from the perspective of "what do we need to do to get this into a place where we can call it 'GA'?" (by which I mean, unlikely to change in incompatible ways and reliable/polished enough for production use). And per offline discussion, I understand that this is probably not something we will say is production-quality, etc. right off the bat.

The question behind my question is: How can we make the HTTP API accessible enough that people will actually use it and give us feedback? If that question is already answered / there's another way to answer it, I agree we don't have to do OpenAPI to start with.

Reply:
> "what do we need to do to get this into a place where we can call it 'GA'?"

In the interest of incremental implementation, I don't think that logic needs to be applied to this proposal or the initial implementation.

> How can we make the HTTP API accessible enough that people will actually use it and give us feedback?

To be honest, I don't really think of this in that way. It's mostly too late for feedback as our API already exists and is GA. And I don't think we need to heavily promote just adding a second protocol to our existing API. This is an existing API, not a new API.

Reply:
> > "what do we need to do to get this into a place where we can call it 'GA'?"
> 
> In the interest of incremental implementation, I don't think that logic needs to be applied to this proposal or the initial implementation.

Yes, we agree here.

> > How can we make the HTTP API accessible enough that people will actually use it and give us feedback?
> 
> To be honest, I don't really think of this in that way. It's mostly too late for feedback as our API already exists and is GA. The only feedback we can get is URL path and that is mostly subjective and dependent upon what exists. And I don't think we need to heavily promote just adding a second protocol to our existing API. This is an existing API, not a new API. This isn't like developing a new product IMO, just a tiny feature on an existing product.

Ah, so my point was more about accessibility than feedback per se. (Though I do think we will want feedback, even if only so that we can consider it if/when we ever decide to do something more RESTful.)

More specifically: If not via OpenAPI, then how will people figure out how gRPC maps to HTTP—e.g. what URL paths are available to use and what those paths accept and return?

Reply:
> If not via OpenAPI, then how will people figure out how gRPC maps to HTTP

OpenAPI can be added later. How do people today figure out what the RPC calls are? If the answer is "look at the proto" and that is good enough for us then the same answer applies to HTTP. Having said that, I would expect docs could be auto-generated. This is what many gRPC+HTTP companies do today, mostly following https://aip.dev/ like https://cloud.google.com/service-infrastructure/docs/apis (which are gRPC-to-HTTP APIs like this proposal proposes IIRC).